engine/graphics/src/null/graphics_null.cpp

```diff
@@ -797,16 +817,70 @@ namespace dmGraphics
         return true;
     }
 
+    static uint32_t NullGetAttributeCount(HProgram prog)
+    {
+        return ((Program*) prog)->m_Attributes.Size();
+    }
+
+    static uint32_t GetElementCount(Type type)
+    {
+        switch(type)
+        {
+            case TYPE_BYTE:
+            case TYPE_UNSIGNED_BYTE:
+            case TYPE_SHORT:
+            case TYPE_UNSIGNED_SHORT:
+            case TYPE_INT:
+            case TYPE_UNSIGNED_INT:
+            case TYPE_FLOAT:
+                return 1;
+                break;
+            case TYPE_FLOAT_VEC4:       return 4;
+            case TYPE_FLOAT_MAT4:       return 16;
+            case TYPE_SAMPLER_2D:       return 1;
+            case TYPE_SAMPLER_CUBE:     return 1;
+            case TYPE_SAMPLER_2D_ARRAY: return 1;
+            case TYPE_FLOAT_VEC2:       return 2;
+            case TYPE_FLOAT_VEC3:       return 3;
+            case TYPE_FLOAT_MAT2:       return 4;
+            case TYPE_FLOAT_MAT3:       return 9;
+        }
+        assert(0);
+        return 0;
+    }
+
+    static void NullGetAttribute(HProgram prog, uint32_t index, dmhash_t* name_hash, Type* type, uint32_t* element_count, uint32_t* num_values, int32_t* location)
+    {
+        Program* program       = (Program*) prog;
+        ShaderBinding& binding = program->m_Attributes[index];
+        *name_hash             = dmHashString64(binding.m_Name);
+        *type                  = binding.m_Type;
+        *element_count         = GetElementCount(binding.m_Type);
+        *num_values            = binding.m_Size;
+        *location              = binding.m_Index;
+    }
+
     static uint32_t NullGetUniformCount(HProgram prog)
     {
         return ((Program*)prog)->m_Uniforms.Size();
     }
 
+    static uint32_t NullGetVertexDeclarationStride(HVertexDeclaration vertex_declaration)
+    {
+        uint32_t stride = 0;
+        for (int i = 0; i < vertex_declaration->m_StreamDeclaration.m_StreamCount; ++i)
+        {
+            VertexStream& stream = vertex_declaration->m_StreamDeclaration.m_Streams[i];
+            stride += GetTypeSize(stream.m_Type) * stream.m_Size;
+        }
+        return stride;
+    }
```

Comment:
Here we assume any vertex declaration will always be tightly packed.
Will it have any performance impact on the GPU, when attributes may end up "unaligned" with respect to their type? (E.g. a float starting at byte offset 1, as opposed to a multiple 4 byte alignment)

Also, If I were to use a C++ struct as-is, from a 2rd party library, I cannot use our helper functions, but need to roll my own?
I guess it should be a small change later on to allow for offsets (like in our buffer impl: https://github.com/defold/defold/blob/dev/engine/dlib/src/dlib/buffer.cpp#L163-L187)

Reply:
True. I'm not sure we will notice the impact for the null adapter since we typically don't access individual 'vertices' afaik. OpenGL and Vulkan has alignment requirements, but I think we adhere to those currently. I'll make a note about it at least.